ravenframework/TSA/MarkovAR.py

```diff
@@ -0,0 +1,333 @@
+    # remaining in the state, which is the state with the greatest diagonal element in the transition
+    # matrix.
+    initState = np.diag(transition).argmax()
+    # We'll use a burn-in period to address (2). In ravenframework/SupervisedLearning/ARMA.py,
+    # the burn-in period is set to twice the maximum AR or MA order. We'll do the same here.
+    burn = 2 * order
+    states = np.zeros(size + burn, dtype=int)
+    states[0] = initState
+    # To transition from one state to another, we need to perform a weighted random choice according
+    # to the transition probabilities. We can do this by creating a cumulative distribution function
+    # for the transition probabilities of each state (cumulative sum of transition probabilities) and
+    # using numpy's searchsorted function to map a uniform random number from U(0, 1) to a state number.
+    transCDFs = np.cumsum(transition, axis=1)
+    rands = engine.uniform(size=size+burn-1)
+    for t in range(1, size + burn):
+      states[t] = np.searchsorted(transCDFs[states[t - 1]], rands[t - 1])
+
+    # Start building up the synthetic signal, starting with the noise
+    synth = engine.normal(loc=0., scale=noiseScale[states], size=size+burn)
+
+    # Build up the AR terms one time step at a time
+    for t in range(order, size + burn):
+      synth[t] += np.sum(ar[states[t]][::-1] * synth[t-order:t])
+
+    # Finally add in the state constants
+    synth += consts[states]
+
+    # There's a chance that the model that has been fit is not stationary, leading to potential divergence
+    # of the synthetic signal. We can check for this by looking for non-finite values in the synthetic
+    # signal.
+    if not np.all(np.isfinite(synth)):
+      raise RuntimeError('Synthetic signal contains non-finite values!')
+
+    # Remove burn-in period
+    synth = synth[burn:]
+    states = states[burn:]
+
+    return synth, states
+
+  def getResidual(self, initial, params, pivot, settings):
+    """
+      Removes trained signal from data and find residual
+      @ In, initial, np.array, original signal shaped [pivotValues, targets], targets MUST be in
+                               same order as self.target
+      @ In, params, dict, training parameters as from self.characterize
+      @ In, pivot, np.array, time-like array values
+      @ In, settings, dict, additional settings specific to algorithm
+      @ Out, residual, np.array, reduced signal shaped [pivotValues, targets]
+    """
+    residual = initial.copy()
+    for tg, (target, data) in enumerate(params.items()):
+      # The model fit results contain the fit residuals, so we can just use those.
+      # The length of the residuals array is shorter than the length of the original signal by the
+      # AR order (not really sure why), so we need to pad the end of the residual array with zeros.
+      modelResid = data['MarkovAR']['model'].resid
+      residual[:, tg] = np.r_[modelResid, np.zeros(len(residual) - len(modelResid))]
```

Comment:
Yeah, the residual should be zero mean, but of course there's no guarantee of that. I initially had it as `np.nan` like a masked value, but I didn't want that causing problems downstream for no real reason. I'd say to just leave it, but I'll add a comment in the code that it might be worth revisiting, especially if these residuals are used by another algorithm in the future.